source/android/logging.c

```diff
@@ -0,0 +1,187 @@
+    size_t next_index = current_index + amount;
+    if (next_index > maximum) {
+        next_index = maximum;
+    }
+
+    return next_index;
+}
+
+/* Override this for Android, as time and log level are taken care of by logcat */
+static int s_logcat_format(struct logcat_format_data *formatting_data, va_list args) {
+    size_t current_index = 0;
+
+    if (formatting_data->total_length == 0) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    /*
+     * Use this length for all but the last write, so we guarantee room for the newline even if we get truncated
+     */
+    size_t fake_total_length = formatting_data->total_length - 1;
+
+    if (current_index < fake_total_length) {
+        /*
+         * Add thread id and user content separator (" - ")
+         */
+        aws_thread_id_t current_thread_id = aws_thread_current_thread_id();
+        char thread_id[AWS_THREAD_ID_T_REPR_BUFSZ];
+        if (aws_thread_id_t_to_string(current_thread_id, thread_id, AWS_THREAD_ID_T_REPR_BUFSZ)) {
+            return AWS_OP_ERR;
+        }
+        int thread_id_written =
+            snprintf(formatting_data->buffer + current_index, fake_total_length - current_index, "] [%s] ", thread_id);
+        if (thread_id_written < 0) {
+            return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        }
+        current_index = s_advance_and_clamp_index(current_index, thread_id_written, fake_total_length);
+    }
+
+    if (current_index < fake_total_length) {
+        /* output subject name */
+        if (formatting_data->subject) {
+            int subject_written = snprintf(
+                formatting_data->buffer + current_index,
+                fake_total_length - current_index,
+                "[%s]",
+                formatting_data->subject);
+
+            if (subject_written < 0) {
+                return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+            }
+
+            current_index = s_advance_and_clamp_index(current_index, subject_written, fake_total_length);
+        }
+    }
+
+    if (current_index < fake_total_length) {
+        int separator_written =
+            snprintf(formatting_data->buffer + current_index, fake_total_length - current_index, " - ");
+        current_index = s_advance_and_clamp_index(current_index, separator_written, fake_total_length);
+    }
+
+    if (current_index < fake_total_length) {
+        /*
+         * Now write the actual data requested by the user
+         */
+
+        int written_count = vsnprintf(
+            formatting_data->buffer + current_index, fake_total_length - current_index, formatting_data->format, args);
+
+        if (written_count < 0) {
+            return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        }
+
+        current_index = s_advance_and_clamp_index(current_index, written_count, fake_total_length);
+    }
+
+    /*
+     * End with a newline.
+     */
+    int newline_written_count =
```

Comment:
Do we ever log stack traces?  We may want to add a truncation indicator given the 1k limit.  Or maybe bump it up to 2-4k?

Reply:
good call. 4K it is. Will fix in follow-up.